keps/sig-node/20190604-rootless.md

```diff
@@ -0,0 +1,213 @@
+- [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Testing](#testing)
+- [History](#history)
+<!-- /toc -->
+
+## Summary
+
+Allow running the entire Kubernetes components (`kubelet`, CRI, OCI, CNI, and all `kube-*`) as a non-root user on the host.
+
+See [FOSDEM 2019 talk "Rootless Kubernetes"](https://www.slideshare.net/AkihiroSuda/rootless-kubernetes) by [@AkihiroSuda](https://github.com/AkihiroSuda) and [@giuseppe](https://github.com/giuseppe) for the overview of Rootless mode.
+See [Usernetes](https://github.com/rootless-containers/usernetes) for the POC.
+
+Rootless mode has been already adopted by k3s, though it doesn't support setting resource limitation as it lacks support for cgroup.
+
+In this KEP, the resource limitation is planned to be supported using cgroup v2, which officially supports delegating its control to a non-root user.
+
+## Motivation
+
+* Protect the host from potential container-breakout vulnerabilities. This is the main movation.
+* Allow users of shared machines (especially HPC) to run Kubernetes without the risk of accidentally breaking their colleagues' environments.
+  Not recommended for real multi-tenancy where the users cannot be trusted.
+* Safe [`kind`](https://github.com/kubernetes-sigs/kind).
+* Safe Kubernetes-on-Kubernetes, to isolate workloads more strictly than Kubernetes API namespaces.
+
+### FAQ: why not use admission controllers?
+Admission controllers like PSP can restrict containers to use extra security options like SELinux, gVisor, and potentially [Node-level UserNS](https://github.com/kubernetes/enhancements/issues/127) in future.
+
+However, these are not efficient to mitigate vulnerabilities of the node components themselves (kubelet, CRI, OCI...).
+
+e.g.
+
+- [CVE-2017-1002102](https://nvd.nist.gov/vuln/detail/CVE-2017-1002102): kubelet could delete files on the host during syncing secret/configMap/downwardAPI volumes
+- [CVE-2019-11245](https://nvd.nist.gov/vuln/detail/CVE-2019-11245): Dockerfile USER instruction was ignored
+- [CVE-2018-11235](https://nvd.nist.gov/vuln/detail/CVE-2018-11235): kubelet could execute an arbitrary command as the root via gitRepo volumes
+- Potential image extraction [zip-slip](https://snyk.io/research/zip-slip-vulnerability) vulnerabilities in CRI runtimes. Both containerd and CRI-O are working on implementing supports for new archive formats like zstd, imgcrypt, and stargz. Potentially these implementations have such vulnerabilities.
+- And lots of CRI/OCI vulnerabilities in the past.
+
+## Goals
+* Step 1: Allow `kubelet` and `kube-proxy` being executed inside user namespaces create by a non-root user [**mergable**]
+  * See [User namespaces](#user-namespaces) and https://github.com/kubernetes/kubernetes/pull/78635
+* Step 2: Allow `kubelet` to configure cgroup v2 via a user-instance of systemd [**Waiting for cgroup2 KEP to settle down**]
```

Comment:
yeah, I just suggested to have cgroup v2 being support by K8s first, then start step 2; otherwise, kubelet is incomplete.

Reply:
cgroup2 is ready now